src/lib/work_selector/offset.ml

```diff
@@ -0,0 +1,35 @@
+open Core_kernel
+
+module Make
+    (Inputs : Intf.Inputs_intf)
+    (Lib : Intf.Lib_intf with module Inputs := Inputs) =
+struct
+  (* Offset will be generated randomly at each execution, and will stay
+     constant for the duration of the execution *)
+  let offset =
+    (* Self init makes Random.int produce a new number at each execution *)
+    Random.self_init () ;
+    (* Having the largest possible value here helps to approach a uniform
+       probability when applying modulo reduction *)
+    Random.int Int.max_value
+
+  let work ~snark_pool ~fee ~logger (state : Lib.State.t) =
+    Lib.State.remove_old_assignments state ~logger ;
+    let unseen_jobs = Lib.State.all_unseen_works state in
+    match Lib.get_expensive_work ~snark_pool ~fee unseen_jobs with
+    | [] ->
+        None
+    | expensive_work ->
+        let n = List.length expensive_work in
+        let x = List.nth_exn expensive_work (offset mod n) in
```

Comment:
i'm not sure how to make this uniform yet, I will see if I can come up with something and report back.

Reply:
The natural way to do it (and a more faithful encoding of the intention anyway) is to modfy `all_unseen_works` to take an optional rotation parameter:
```ocaml
    let all_unseen_works ?offset t =
      let available_jobs =
        match offset with
        | None -> t.available_jobs
        | Some offset -> (
            let (prefix, suffix) = List.split_n t.available_jobs offset in
            suffix @ prefix)
      in
      O1trace.sync_thread "work_lib_all_unseen_works" (fun () ->
          List.filter available_jobs ~f:(fun js ->
              not
              @@ Map.mem t.jobs_seen (One_or_two.map ~f:Work_spec.statement js) ) )
```
Of course, this means that you need to make sure that `offset` is in the range of values that `List.length t.available_jobs` may take (I think this is `0 <= offset < (Int.pow 2 constraint_constants.transaction_capacity_log_2) * constraint_constants.work_delay`, but it's probably best to check with @nholland94), so you may need to route that parameter through from somewhere.

Of course, the current implementation -- and correspondingly this proposal -- is wasteful because it creates a full list of items and evaluates the predicate over it just to select a single item.

Reply:
See I was so skeptical that the zero logic in sequential did not apply to an arbitrary offset!

Separately, I thought an optimistic upper bound for snark worker pool size was (max number of proofs * max number of trees) i.e `((Int.pow 2 constraint_constants.transaction_capacity_log_2 + 1)  - 1)* ((constraint_constants.transaction_capacity_log_2 + 1) * (work_delay + 1) + 1)`

Perhaps uniformity could be achieved with this version of the `all_unseen_works` function and a randomly chosen offset from 0 to n (size of work list). This choice can then be reused until it is larger than n. When this happens we can just repeat the selection process. Imo this is still rand + seq despite the fact the offset changes based on the work list size. It should be possible to just index into the offset in O(offset) instead of O(n) with this solution as well.

Reply:
I’m not sure to see (in terms of final result & global performance) how this implementation of `all_unseen_works` differs from the current version with an offset in the right bounds (which I would prefer in terms of code cleanliness), but it is possible that I’m missing something.

Reply:
this is how i modeled the difference for myself:

```python
In [56]: def test_get_work():
    ...:     expensive_work = [i for i in range(20)]
    ...:     # choose n with uniform probability
    ...:     n = randint(0, len(expensive_work))
    ...:     for i in range(len(expensive_work) - n):
    ...:         print(f"n: {n}", expensive_work.pop(n))
    ...:     # re choose n at this point on the smaller expensive_work with uniform probability i.e recurse
    ...:
```
```
In [58]: test_get_work()
n: 8 8
n: 8 9
n: 8 10
n: 8 11
n: 8 12
n: 8 13
n: 8 14
n: 8 15
n: 8 16
n: 8 17
n: 8 18
n: 8 19```

vs

```python 
In [57]: def test_get_work_mod():
    ...:     expensive_work = [i for i in range(20)]
    ...:     n = randint(0, 2 ** 30)
    ...:     print(f"n: {n}")
    ...:     while expensive_work:
    ...:         print(f" n mod len: {n % len(expensive_work)}", expensive_work.pop(n % len(expensive_work)))
    ...:

```

```
test_get_work_mod()
n: 1028505273
 n mod len: 13 13
 n mod len: 9 9
 n mod len: 15 17
 n mod len: 3 3
 n mod len: 9 11
 n mod len: 3 4
 n mod len: 5 7
 n mod len: 3 5
 n mod len: 9 16
 n mod len: 4 8
 n mod len: 3 6
 n mod len: 6 15
 n mod len: 1 1
 n mod len: 5 18
 n mod len: 3 12
 n mod len: 3 14
 n mod len: 1 2
 n mod len: 0 0
 n mod len: 1 19
 n mod len: 0 10
```

the first implementation seems truer to the sequential mode that always pulls from offset 0